Reifier.fs

```diff
@@ -12,54 +12,249 @@ open Starling.Core.Model
+       and which are (n)ot.  We assume func names are unique and all func
+       references are well-formed (correct set of parameters), which should have
+       been checked earlier. *)
+    let ifuncs, nfuncs =
+        protos
+        |> FuncDefiner.toSeq
+        |> Seq.fold
+            (fun (is, ns) (func, pinfo) ->
+                if pinfo.IsIterated
+                then (Set.add func.Name is, ns)
+                else (is, Set.add func.Name ns))
+            (Set.empty, Set.empty)
+
+    (* Now, go through the multiset's funcs.  Since the multiset may contain
+       n copies of a k-iterated func, we normalise to one copy of a k*n-iterated
+       func to get rid of the outer iterator.  What we then do depends on
+       whether the func is iterated or not: if it isn't, we just emit it;
+       otherwise, we add it to an equivalence class based on the func name so
+       we can do the case-split expansion. *)
+    let expandFuncToReify
+      (ns, ic : Map<string, IteratedGFunc<Sym<MarkedVar>> list>) func n =
+        let norm = TermGen.normalise func n
+        let nname = norm.Func.Item.Name
+        if ifuncs.Contains nname
+        then (ns, Map.add nname (norm::ic.[nname]) ic)
+        else (norm::ns, ic)
+
+    // Have to make sure each class exists first; else exceptions happen.
+    let icempty = ifuncs |> Seq.map (fun name -> (name, [])) |> Map.ofSeq
+    let nlist, iclasses = Multiset.fold expandFuncToReify ([], icempty) view
+
+    (* Now, go through the equivalence classes to calculate their case-split
+       expansion.  We do this by working out every single possible set of
+       equalities between the funcs in the class: say, for
+
+       G1->A(a)[i] * G2->A(b)[j] * G3->A(c)[k]
+
+       we have the cases (), (a=b), (a=c), (b=c), (a=b && a=c).  This turns out
+       to be the powerset of the class, less the empty set, where each element
+       of the powerset denotes equality between the members' parameters.
+       We don't need the map keys at this stage. *)
+    let ipsets : (IteratedGFunc<Sym<MarkedVar>> list) seq =
+        iclasses |> Map.toSeq |> Seq.map (snd >> powerset) |> Seq.concat
+
+    (* Finally, we need to convert those equivalence powersets into a list of
+       funcs. *)
+    let mergeEqualitySet xs =
+        function
+        // Trivial cases first.
+        | [] -> xs
+        | [func] -> func::xs
+        (* Now we have a set of func G1->A(a)[i], G2->A(b)[j], G3->A(c)[k]...
+           first, calculate the new guard G1^G2^G3^a=b^a=c^... *)
+        | gfuncs ->
+            let guards, funcs =
+                gfuncs
+                |> List.map iterGFuncTuple
+                |> List.unzip
+            let equalities = paramEqualities funcs
+            let nguard = mkAnd (guards @ equalities)
+
+            // And the new iterator i+j+k+...
+            let iter =
+                funcs
+                |> List.map (fun f -> f.Iterator)
+                |> List.fold mkAdd2 (AInt 0L)
```

Comment:
I'm adding `mkAdd` for this purpose, though the implementation will be slightly bad for now